### PR TITLE
Improved login page layout and control

### PR DESCRIPTION
**Pull request by robertdown**
Refactor the login twig template to several partials and layouts to improve template reuse by module developers. Also allows a no-code solution to customizing primary, secondary, and small logos at the site-level.

# Overriding The Default Logos

## Primary Logo
* Place an image file in `site/<site_id>/images/logos/core/login/primary/` which **must** be named `logo`

## Secondary Logo
* Place an image file in `site/<site_id>/images/logos/core/login/secondary/` which **must** be named `logo`
* Make sure to enable "Show Secondary Logo On Login Page" option in Admin -> Config -> Login Page

## Small Logo 1
* Place an image file in `site/<site_id>/images/logos/core/login/small_logo_1/` which **must** be named `logo`
* Make sure to enable "Show Mini Logo 1" option in Admin -> Config -> Login Page

## Small Logo 2
* Place an image file in `site/<site_id>/images/logos/core/login/small_logo_2/` which **must** be named `logo`
* Make sure to enable "Show Mini Logo 2" option in Admin -> Config -> Login Page

If the end user does not override an image, the default logo will be rendered from `public` ensuring continuity during upgrades.

There are now several different layouts, the box (status quo) which can render logos on top, left, or right. There is now a horizontal band (logos can go on left or right) and vertical band.

# Overriding and Extending the Login Page
Developers can now access the login page through various twig partials and and Event. See `templates/login/partials` for all the partials that can be used. Developers will `{% extend "login/base.html.twig" %}` in their custom layout and then define new `{% block %}` tags to override or extend the content. Additionally, add a listener to TemplatePageEvent::RENDER_EVENT with `pageName` set to `login/login.php` to hook into the rendering process. Override the Twig Template file with `setTwigTemplate()` and manage the Twig variables with `setTwigVariables()` (Which will merge the variables) or `emptyTwigVariables()` which will allow you to completely remove all variables - this is likely **not** what you want to do in most cases.

### CODE REVIEW

**Comment by robertdown** [comment]
Todo

- [ ] Trigger an event for Twig rendering, allowing new template variables to be injected
- [ ] Commit the new file structure in sites/default
- [ ] Commit the new default logo in public

File: composer.json [hunk 1]
```diff
@@ -72,7 +72,8 @@
         "twig/twig": "3.4.3",
         "vlucas/phpdotenv": "5.5.0",
         "waryway/php-traits-library": "1.0.4",
-        "yubico/u2flib-server": "1.0.2"
+        "yubico/u2flib-server": "1.0.2",
+        "symfony/finder": "^5.4"
```
Comment (bradymiller) [inline composer.json:76]: pin the full version and put above in symfony

File: interface/login/login.php [hunk 2]
```diff
@@ -40,6 +41,59 @@
...
+$baseSmallLogoTwoDir = $GLOBALS['OE_SITE_DIR'] . "/images/logos/core/login/small_logo_2";
+$smallLogoTwoFinder->files()->in($baseSmallLogoTwoDir)->name("logo.*");
+
+if ($smallLogoTwoFinder->hasResults()) {
+    foreach ($smallLogoTwoFinder as $f) {
+        $smallLogoTwoPath = $GLOBALS['OE_SITE_WEBROOT'] . "/images/logos/core/login/small_logo_2/" . $f->getFileName();
+    }
+} else {
+    $smallLogoTwoPath = $GLOBALS['OE_SITE_WEBROOT'] . "images/logo_2.png";
+}
```
Comment (bradymiller) [inline interface/login/login.php:93]: Would be ideal to have a Logo class that encapsulated above, so could then grab the paths for each logo with a one liner that could then be used in other places (for example the patient portal and the oauth2 client page thingy).

File: library/globals.inc.php [hunk 3]
```diff
@@ -538,6 +527,30 @@ function gblTimeZones()
             '1',
             xl('Display the Donations link on the About page'),
         ],
+    ],
+
+    // Login Page
+    'Login Page' => [
+        'login_page_layout' => array(
```
Comment (bradymiller) [inline library/globals.inc.php:534]: What will happen to upgraders that have center, left, right for this setting. If odd things happen may make sense to change the id of this setting (and then ignore the old setting id).

Comment (bradymiller) [reply library/globals.inc.php:534]: I just noted below that you addressed this in the sql upgrade script, so disregard above post :)

File: sql/7_0_0-to-7_0_1_upgrade.sql [hunk 4]
```diff
@@ -286,3 +286,19 @@ ALTER TABLE `form_questionnaire_assessments` CHANGE `user` `user` VARCHAR(255) N
...
+UPDATE `globals` SET `gl_value` = 'login/layouts/horizontal_box_left_logo.html.twig' WHERE `gl_name` = 'login_page_layout' AND `gl_value` = 'left';
+#EndIf
+
+#IfRow2D globals gl_name login_page_layout gl_value right
+UPDATE `globals` SET `gl_value` = 'login/layouts/horizontal_band_right_logo.html.twig' WHERE `gl_name` = 'login_page_layout' AND `gl_value` = 'right';
+#EndIf
+
+#IfNotRow globals gl_name secondary_logo_position
+INSERT INTO `globals` gl_name='secondary_logo_position', gl_value='second';
+#EndIf
```
Comment (bradymiller) [inline sql/7_0_0-to-7_0_1_upgrade.sql:304]: remove above 3 lines. new globals are automatically added with default in globals.inc.php during install/upgrade, so no need to put here.

File: templates/login/partials/html/primary_logo.html.twig [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+@subpackage Login
+@author Robert Down <robertdown@live.com
+@copyright Copyright (c) 2023 Robert Down <robertdown@live.com>
+@copyright Copyright (c) 2023 Providence Healthtech
+@license https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+#}
+{% set secondaryLogoPosition = (secondaryLogoPosition == 'first') ? 'flex-column-reverse' : ''  %}
+<div class="{{ logo_container_class_list|default(['d-flex', 'flex-column', 'justify-content-between', 'align-items-center'])|join(' ') }} {{ logo_flex_direction|default([])|join(' ') }} {{ secondaryLogoPosition }}">
+    <div class="{{ primary_logo_container_class_list|default(['flex-fill'])|join(' ') }}">
+        <img src="{{ primaryLogo }}" class="img-fluid" alt="">
```
Comment (bradymiller) [inline templates/login/partials/html/primary_logo.html.twig:14]: attr

File: templates/login/partials/html/primary_logo.html.twig [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    </div>
+    {% if displaySecondaryLogo and secondaryLogo %}
+        <div class="{{ logo_divider_container_class_list|default(['px-3', 'flex-fill'])|join(' ') }}">
+            <i class="fas fa-plus"></i>
+        </div>
+        <div class="{{ secondary_logo_container_class_list|default('flex-fill')|join(' ') }}">
+            <img src="{{ secondaryLogo }}" alt="" class="img-fluid">
```
Comment (bradymiller) [inline templates/login/partials/html/primary_logo.html.twig:21]: attr

File: templates/login/partials/html/small_logo.html.twig [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+<div class="{{ small_logo_container_class_list|default(['d-flex', 'justify-content-around'])|join(' ') }}">
+{% if displaySmallLogo == 1 %}
+    <img src="{{ smallLogoOne }}" alt="{{ 'Small Logo 1'|xlt }}" class="img-fluid">
+{% endif %}
+{% if displaySmallLogo == 2 %}
+    <img src="{{ smallLogoTwo }}" alt="{{ 'Small Logo 2'|xlt }}" class="img-fluid">
+{% endif %}
+{% if displaySmallLogo == 3 %}
+    <img src="{{ smallLogoOne }}" alt="{{ 'Small Logo 1'|xlt }}" class="img-fluid">
+    <img src="{{ smallLogoTwo }}" alt="{{ 'Small Logo 2'|xlt }}" class="img-fluid">
```
Comment (bradymiller) [inline templates/login/partials/html/small_logo.html.twig:22]: attr smallLogoOne/Two in 4 lines above

**Comment by bradymiller** [comment]
@robertdown , neat stuff!

File: interface/login/login.php (hunk 2 above)
Comment (robertdown) [reply interface/login/login.php]: Done

File: sql/7_0_0-to-7_0_1_upgrade.sql (hunk 4 above)
Comment (robertdown) [reply sql/7_0_0-to-7_0_1_upgrade.sql]: Done

**Comment by robertdown** [comment]
@adunsulag can you review the TemplatePageEvent changes to make sure they are copacetic with that Event more broadly?

**Comment by robertdown** [comment]
Thanks @bradymiller for helping me solve the failing tests! There was a version mismatch between composer.json and the lock file. Fixed now. I believe this PR is ready, once you approve I'll push the Squash and Merge Button 🙂

File: sites/default/images/logos/core/login/primary/README [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# Primary Login Logo
```
Comment (bradymiller) [inline sites/default/images/logos/core/login/primary/README:1]: just to confirm that this mechanism will not break anything if this directory does not exist (for example, upgraders will not have this directory).

File: src/Services/LogoService.php [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+ * @copyright   Copyright (c) 2023 Robert Down
+ * @copyright   Providence Healthtech
+ * @license     https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+
+namespace OpenEMR\Services;
+
+use Symfony\Component\Finder\Finder;
+
+class LogoService
```
Comment (bradymiller) [inline src/Services/LogoService.php:20]: dance-bear

**Comment by bradymiller** [comment]
@robertdown , Code looks good to me (i didn't test it). As long as things don't break if the directories in the sites directory do not exist, then good to go :)

**Comment by robertdown** [comment]
Let me do one more check.

**Comment by robertdown** [comment]
Update - if the folder structure does not exist in `sites` the script throws an exception. Working on a solution now.

**Comment by robertdown** [comment]
Ok, I pulled in the Symfony filesystem component and added a condition to check if the sites directory exists, if it does I add the sites directory to the array of paths to search. It is simple and will ensure upgrades will work without breaking. This truly only manages upgrades as new installs will have the directory structure in `sites` built-in by default.

Just throwing this out there.... If we could execute some code during the upgrade process, we could alleviate this code by just creating the structure. Convo for another day I suppose 😄

**Comment by bradymiller** [comment]
2DV(1)

**Pull request merged**